policy / validation: CPFP fee bumping within packages

glozow opened the pull request:
Part of #22290, aka [Package Mempool Accept](https://gist.github.com/glozow/dc4e9d5c5b14ade7cdfac40f43adb18a).

This enables CPFP fee bumping in child-with-unconfirmed-parents packages by introducing [package feerate](https://gist.github.com/glozow/dc4e9d5c5b14ade7cdfac40f43adb18a#fee-related-checks-use-package-feerate) (total modified fees divided by total virtual size) and using it in place of individual feerate. We also always [validate individual transactions first](https://gist.github.com/glozow/dc4e9d5c5b14ade7cdfac40f43adb18a#always-try-individual-submission-first) to avoid incentive-incompatible policies like "parents pay for children" or "siblings pay for siblings" behavior.

glozow committed:
[packages/policy] use package feerate in package validation

This allows CPFP within a package prior to submission to mempool.

glozow committed:
[docs] package feerate

glozow committed:
[validation] try individual validation before package validation

This avoids "parents pay for children" and "siblings pay for siblings"
behavior, since package feerate is calculated with totals and is
topology-unaware.

It also ensures that package validation never causes us to reject a
transaction that we would have otherwise accepted in single-tx
validation.

glozow committed:
[unit test] package feerate and package cpfp

mzumsande reviewed (commented):
Concept ACK

left a few comments below

LarryRuane commented on doc/policy/packages.md:
```diff
@@ -57,3 +57,40 @@ test accepts):
+used for the transactions that failed.
+
+*Rationale*: Packages are intended for incentive-compatible fee-bumping. That is, transaction A is a
+"legitimate" fee-bump transaction B only if A is a descendant of B and has a higher feerate.
```
nit, most examples seem to designate a parent as A and its child as B (which I like, because that corresponds to time-order, first A, then B), but this seems to be reversed? I believe a few lines below, the convention I'm suggesting is followed, (A is parent, B is child). Also, should "transaction B" be "transaction for B"?

LarryRuane commented on doc/policy/packages.md:
```diff
@@ -57,3 +57,40 @@ test accepts):
+
+### Package Fees and Feerate
+
+*Package Feerate* is the total modified fees divided by the total virtual size of all transactions
```
This is okay, but when I first read this, I had no idea what "modified" meant (only learning later that the `prioritisetransaction` RPC does the modifying); perhaps "modified" can be dropped? I think it's implicit. Or maybe briefly mention the RPC here.

glozow replied:
Added explanation

LarryRuane commented on doc/policy/packages.md:
```diff
@@ -57,3 +57,40 @@ test accepts):
+### Package Fees and Feerate
+
+*Package Feerate* is the total modified fees divided by the total virtual size of all transactions
+in the package after deduplication.
```
It's unclear to me what "deduplication" means here. I thought a package, by definition, contains no duplicate transactions.

michaelfolkson commented on src/test/txpackage_tests.cpp:
```diff
@@ -557,6 +577,182 @@ BOOST_FIXTURE_TEST_CASE(package_witness_swap_tests, TestChain100Setup)
+    }
+
+    // Package feerate is calculated without topology in mind; it's just aggregating fees and sizes.
+    // However, this should not allow parents to pay for children. Each transaction should be
```
"However, this should not allow parents to pay for children"

This reads to me like "parents must not pay for children". When it is more (to my understanding), parents must be validated first before considering children. If the parent(s) is accepted into the mempool first a low fee child can later join them as part of a sufficient fee package but a low fee child must not prevent a sufficient fee parent from being accepted into the mempool. That's right yeah?

michaelfolkson replied:
Or maybe because in the process of a child (subsidized by a parent) being accepted into the mempool the parent could be mined on its own and then the child would need to be kicked out the mempool?

glozow replied:
We shouldn't accept the child into the mempool if it doesn't meet the minimum feerate.

michaelfolkson replied:
Ah its a miner's economic interest argument. Ok gotcha, makes sense. Thanks

michaelfolkson commented:
Concept ACK

Built on MacOS Big Sur, ran unit tests (diagrams from [PR review club](https://bitcoincore.reviews/24152) are helpful here illustrating each unit test). Ideally I'd play around with these but just read through and ran them today. Didn't run the fuzz tests but the fuzzer seems to be failing on the CI.

I asked about DoS vectors for CPFP in today's PR review club as that seems the greatest hurdle to get over for updated RBF rules and package RBF but seem to be less of a concern with CPFP (?). Not sure whether the CPFP carve out rule needs to be considered when CPFPing within a package?

Great to see this series of PRs progress.

ariard commented on doc/policy/packages.md:
```diff
@@ -57,3 +57,40 @@ test accepts):
+A to be mined, B's fees cannot help A. In a child-with-parents package, simply excluding any parent
+transactions that meet feerate requirements individually is sufficient to ensure this.
+
+*Rationale*: For backwards-compatibility of transaction relay policy, it's important that package
```
i agree as a design principle that backwards-compatibility of transaction relay policy is worthwhile.

That said, we might have in the future to make exceptions to that principle, where a transaction/package might be relayed across non-upgraded nodes though rejected by upgraded nodes. E.g I'm thinking if replace-by-feerate is deployed instead of replace-by-fee, you might have a (2000 sats; 10 sat/vB) transaction replacing a (5000 sats; 5 sat/vB) already-in-mempool one (even if we can envision some transition period with dual-support).

Thus I'm suggesting to lightweight a bit the wording here, "As a principle, backwards-compatibility of transaction relay policy is prioritized", "we prevent non-programmed restriction of policy". Especially as all those questions of policy loosening/tightening and how the project operates them are relatively young, I think it's cautious to be more formal in the phrasing. Thereby avoiding in the future ecosystems users to claim an adamant stability of Core policy rules, like in the zero-conf/full-rbf past polemics.

glozow replied:
Agree, added a sentence at the beginning to this effect.

ariard commented on src/test/txpackage_tests.cpp:
```diff
@@ -557,6 +577,182 @@ BOOST_FIXTURE_TEST_CASE(package_witness_swap_tests, TestChain100Setup)
+
+    // Package feerate includes the modified fees of the transactions.
+    // This means a child with its fee delta from prioritisetransaction can pay for a parent.
+    m_node.mempool->PrioritiseTransaction(tx_child_cheap->GetHash(), 1 * COIN);
```
From checking `prioritisetransaction` documentation, I think it's a valid behavior to subtract fees from a transaction, thus altering `ProcessNewPackage` result. If you think it's a case worthy to cover.

glozow replied:
Good idea! I will add a test for this.

ariard reviewed (commented):
Concept ACK

Sagor2214 commented on doc/policy/packages.md:
```diff
@@ -57,3 +57,40 @@ test accepts):
+used for the transactions that failed.
+
+*Rationale*: Packages are intended for incentive-compatible fee-bumping. That is, transaction A is a
+"legitimate" fee-bump transaction B only if A is a descendant of B and has a higher feerate.
```
Thanks the help

glozow commented on src/validation.cpp:
```diff
@@ -649,6 +656,10 @@ class MemPoolAccept
     CChainState& m_active_chainstate;
 
+    // Used to calculate package feerate.
+    CAmount m_total_modified_fees;
```
Yes good point, changed now. I originally added it as a member variable so we could reuse it in `PackageRBFChecks()` in the future, but we'll cross that bridge when we get there.

glozow commented on src/validation.cpp:
```diff
@@ -1199,6 +1211,21 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
+    if (args.m_package_feerates &&
+        !CheckFeeRate(m_total_vsize, m_total_modified_fees, placeholder_state)) {
+        package_state.Invalid(PackageValidationResult::PCKG_POLICY, "package-fee-too-low");
+        return PackageMempoolAcceptResult(package_state, package_feerate, {});
```
Ooooh good catch

glozow commented:
Thanks @ariard @mzumsande @LarryRuane @michaelfolkson! Addressed your review comments. Clarified the documentation, adjusted the fuzz/tx_pool, and added the test for negative prioritisation.

LarryRuane commented:
Concept ACK

glozow commented:
Rebased

ariard commented on src/validation.cpp:
```diff
@@ -1338,18 +1376,33 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
             // Transaction does not already exist in the mempool.
-            txns_new.push_back(tx);
+            // Try submitting the transaction on its own.
+            const auto single_res = AcceptSingleTransaction(tx, single_args);
```
I think the `TxValidationState` could be inspected here. If the failure cannot be assigned to the feerate check, the package validation could fail early ? Otherwise if you have a package of 25th component and the latest is faultive, we might validate twice the first 24 ones. Once here, the other one in `AcceptMultipleTransactions` and this could be avoid.

Am I correct ?

If yes, I think it's always good to minimize our DoS surface. Though I'm unsure if the difference would be significant between the two implementations.

glozow replied:
Correct - if the failure isn't feerate or missing inputs (i.e. the child, or the parents could depend on each other), then there's nothing package validation can do. I had this in the original version and took it out, but you're right that it's safer to have here. I've added it back in the latest push.

ariard commented on src/validation.cpp:
```diff
@@ -1338,18 +1376,33 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
 
     // Nothing to do if the entire package has already been submitted.
-    if (txns_new.empty()) return PackageMempoolAcceptResult(package_state, std::move(results));
+    // Also quit now if any of the transactions was consensus-invalid because package validation
```
IIUC `txns_new` is empty only if all submitted package transactions are accepted. Not because transactions are consensus-invalid, I'm unsure if the comment accurate with the latest version of the code ?

glozow replied:
Yeah sorry, leftover comment

glozow committed:
[validation] don't package validate if not policy or missing inputs

Package validation policy only differs from individual policy in its
evaluation of feerate. Minimize DoS surface; don't validate all over
again if we know the result will be the same.

glozow commented:
Fixed the fuzzer issue, added a commit for quitting early if transactions fail for reasons other than policy or missing inputs.

glozow commented:
pinging @ariard @darosior @LarryRuane @mzumsande if you have time to take another look? :)

ariard reviewed (commented):
Code Review ACK a7017507

Modifications since last review are typying arguments in `AcceptMultipleTransactions` and `quit_early` new a7017507 commit.

glozow commented:
Rebased

jonatack commented on doc/policy/packages.md:
```diff
@@ -72,3 +72,48 @@ test accepts):
+
+*Rationale*: As a principle, we want to avoid accidentally restricting policy in order to be
+backwards-compatibile for users and applications that rely on p2p transaction relay. Concretely,
+it's important that package validation does not prevent the acceptance of a transaction that would
```
prefer active voice: s/it's important that package validation does not/package validation must not/

glozow replied:
Took all the doc comments 👍

jonatack commented on src/test/txpackage_tests.cpp:
```diff
@@ -568,6 +587,205 @@ BOOST_FIXTURE_TEST_CASE(package_witness_swap_tests, TestChain100Setup)
+{
+    mineBlocks(5);
+    LOCK(cs_main);
+    unsigned int expected_pool_size = m_node.mempool->size();
```
22bf054
- prefer absolute references to cs_main in new code
- use `size_t` type for `size()` (or `auto`), same for line 216
```diff
-    LOCK(cs_main);
-    unsigned int expected_pool_size = m_node.mempool->size();
+    LOCK(::cs_main);
+    size_t expected_pool_size{m_node.mempool->size()};
```

glozow replied:
Done, noted

jonatack commented on src/test/txpackage_tests.cpp:
```diff
@@ -568,6 +587,205 @@ BOOST_FIXTURE_TEST_CASE(package_witness_swap_tests, TestChain100Setup)
+
+        // package feerate should include parent3 and child. It should not include parent1 or parent2_v1.
+        BOOST_CHECK(mixed_result.m_package_feerate != std::nullopt);
+        CFeeRate expected_feerate(1 * COIN, GetVirtualTransactionSize(*ptx_parent3) + GetVirtualTransactionSize(*ptx_mixed_child));
```
22bf054eee6bfc7148bb40150d648b0dcf0db44b nit, this is the only non-const `expected_feerate` in the test and it can be const too
```suggestion
        const CFeeRate expected_feerate(1 * COIN, GetVirtualTransactionSize(*ptx_parent3) + GetVirtualTransactionSize(*ptx_mixed_child));
```

glozow replied:
Done

jonatack reviewed (commented):
WIP review, looked at the doc and starting trying to grok the tests

glozow commented on doc/policy/packages.md:
```diff
@@ -72,3 +72,48 @@ test accepts):
+fees but be paid for by the child.
+
+*Rationale*: This can be thought of as "CPFP within a package," solving the issue of a parent not
+meeting minimum fees on its own. This allows L2 applications to adjust their fees at broadcast time
```
I'll change to "contracting applications." I specifically mean contracting applications that negotiate fees + sign with counterparties ahead of broadcast time

LarryRuane reviewed (commented):
One thing I had to think really hard about (maybe it's just me) is the effect of deduplication. I pictured a package like A <-- B <-- C (A is B's parent, B is C's parent). What happens if (only) B is already in the mempool, so we deduplicate it out of the package? Then the package would consist of A <-- ??? <-- C, which is no longer a valid package.

I think the answer to this is that if B is already in the mempool, then A _must_ also already be in the mempool, so it will be deduplicated out of the package _too_. (This wasn't obvious to me at first.) Or I guess A could be mined, so the package would end up being just C, which is not actually a package (which is okay, evaluate C on its own). In any event, deduplication only removes from the "left" (if you view the transactions payments as flowing left to right as I've done here) -- it can't remove from the middle of this big DAG (well, it can't be a big DAG yet, but will in the future).

One other comment, and again, it may just be me, but it may not be obvious to the uninitiated why package relay is needed at all. Taking the most basic single parent and single child case, if the parent's fee is so low that it drops out of the mempool (or at least many nodes' mempools), then why not just reconstitute the parent with a higher fee and broadcast this new transaction? Why is a child even needed?

I _think_ the answer (and correct me if I'm wrong) is that the parent transaction's broadcaster may not _be able_ to create a new version of the parent transaction. It may have been signed by other parties, and they're not around (or aren't willing) to sign a replacement transaction. In L2 applications especially, this transaction may have been formed months ago.

I'm not sure if I'm suggesting any changes to this document, but you may want to consider trying to get this concept across (if my understanding is correct).

glozow commented on doc/policy/packages.md:
```diff
@@ -72,3 +72,48 @@ test accepts):
+*Rationale*: It would be incorrect to use the fees of transactions that are already in the mempool, as
+we do not want a transaction's fees to be double-counted.
+
+Implementation Note: Transactions within a package are always validated individually first, and
```
I don't agree; we *are* validating the transaction fully. I don't think "feerate-validate" is a commonly used term so could be even more confusing.

glozow commented on doc/policy/packages.md:
```diff
@@ -72,3 +72,48 @@ test accepts):
+we do not want a transaction's fees to be double-counted.
+
+Implementation Note: Transactions within a package are always validated individually first, and
+package validation is used for the transactions that failed. Since package feerate is only
```
Your description is correct at a high level, and is what I intend to communicate to the reader. To be nitpicky, we'll add transactions to the package for *any* policy reason, not just feerate, since `TxValidationResult` doesn't specify whether it's a fee-related or other policy issue.

> I think this algorithm generalizes to an arbitrary package DAG but I'm not completely sure.

While I wish this were true, there are cases where it doesn't work - see [example Q2](https://gist.github.com/glozow/dc4e9d5c5b14ade7cdfac40f43adb18a#always-try-individual-submission-first) in the package mempool accept gist. This is why the package-child-with-unconfirmed-parents restriction is so important.

glozow reviewed (commented):
Thanks @LarryRuane for the thoughtful review.

> One thing I had to think really hard about (maybe it's just me) is the effect of deduplication.  I pictured a package like A <-- B <-- C (A is B's parent, B is C's parent)...

Note that A<-B<-C is already not a valid package with these rules; it would fail `IsChildWithParents`. I'll continue with this example considering just A<-B.

You are correct that if B is already in the mempool, A (and any other ancestors or transactions to the "left") must either be in the mempool or already confirmed. Unfortunately, we cannot query with absolute certainty if a transaction has already confirmed, which is why the rule is it must be a child with all *unconfirmed* parents. The only possibilities for B already being in the mempool are:

1. A in mempool, B in mempool. We deduplicate both; the package is already in.
2. A confirmed, B in mempool. We fail on `package-not-child-with-unconfirmed-parents` because A's inputs were not available. This is correct, since a *confirmed* parent of B was included in the package.

> why not just reconstitute the parent with a higher fee and broadcast this new transaction? Why is a child even needed? I think the answer (and correct me if I'm wrong) is that the parent transaction's broadcaster may not be able to...

You are correct.

LarryRuane reviewed (commented):
Code review, tested ACK dc0b2d3fa375ec0641bd2075a99be0383d1129ee

Suggestions are optional.

Regarding your reply to my earlier [comment](https://github.com/bitcoin/bitcoin/pull/24152#discussion_r832775358):

> there are cases where it [my little strawman algorithm] doesn't work [for the general case]

Thanks, you're right, I see now that this trivial algorithm is definitely wrong. This made me think about the general case more carefully, and I ended up writing a little Python demo to help with my understanding. This in no way affects the current PR (which is excellent as-is), but anyone interested in this topic may find this useful, and it may be helpful in writing a future PR if we want to generalize the allowed package forms:

https://github.com/LarryRuane/bitcoin_package_accept_demo

glozow commented on src/validation.cpp:
```diff
@@ -1231,20 +1254,35 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
+    // Transactions must meet two minimum feerates: the mempool minimum fee and min relay fee.
+    // For transactions consisting of exactly one child and its parents, it suffices to use the
+    // package feerate (total modified fees / total virtual size) to check this requirement.
+    const auto m_total_vsize = std::accumulate(workspaces.cbegin(), workspaces.cend(), int64_t{0},
+        [](int64_t sum, auto& ws) { return sum + ws.m_vsize; });
```
Yeah I disagree. I prefer being able to assign this to a const. I don't see anything wrong with using `std::accumulate`.

glozow commented:
Rebased and addressed @jonatack and @LarryRuane's comments

glozow commented on src/test/txpackage_tests.cpp:
```diff
@@ -399,6 +411,9 @@ BOOST_FIXTURE_TEST_CASE(package_witness_swap_tests, TestChain100Setup)
+        BOOST_CHECK(submit_witness1.m_package_feerate == std::nullopt);
+
         const auto submit_witness2 = ProcessNewPackage(m_node.chainman->ActiveChainstate(), *m_node.mempool,
                                                        {ptx_parent, ptx_child2}, /*test_accept=*/ false);
```
Added

glozow commented on src/validation.cpp:
```diff
@@ -1380,18 +1394,31 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
     for (const auto& [wtxid, mempoolaccept_res] : results) {
         submission_result.m_tx_results.emplace(wtxid, mempoolaccept_res);
     }
+    if (submission_result.m_state.IsValid()) assert(submission_result.m_package_feerate != std::nullopt);
```
Changed all of them to `has_value()`

LarryRuane replied:
If you touch up, this one still could be changed (the test calls have been changed).

ariard commented:
Code Review ACK a7017507

Modifications since last review are extending `ATMPArgs` private ctor extension with `m_package_feerates` and changes in the unit test framework.

fanquake commented:
cc @darosior @t-bast

t-bast reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/24152/commits/157a7bb547abc90a8bbb595cdee8eaac9c3300c9, nice set of unit tests :+1:

instagibbs commented:
Out of scope for now but dropped a message on the gist itself https://gist.github.com/glozow/dc4e9d5c5b14ade7cdfac40f43adb18a

I see no reason for this(possibly wrongly perceived) shortfall to stop incremental improvements

instagibbs commented:
@glozow yes, you're right. Lots of terms, and I misunderstood the OP's scope. Reviewing now.

LarryRuane commented:
> Rebased and addressed @jonatack and @LarryRuane's comments

Thanks, suggestion for future, make the rebase and address-review-comments two separate force-pushes for easier review.

instagibbs reviewed (approved):
code review ACK 157a7bb547abc90a8bbb595cdee8eaac9c3300c9

docs are clear re:design, tests look fairly comprehensive, nice step forward

glozow commented:
> Code Review ACK https://github.com/bitcoin/bitcoin/commit/a70175076a5f85391a089607f6a19a58f173d4a1

@ariard seems you ACKed an old commit? A rebase + small changes since then, should be pretty easy to review `git range-diff a701750...157a7bb`

mzumsande reviewed (commented):
Reviewed again and it looks good to me, just two questions before I ACK:

Could this introduce some weird behavior for the `testmempoolaccept` RPC? That RPC bypasses `AcceptPackage()` and calls `AcceptMultipleTransactions()` directly, which now has logic to calculate a package feerate (and possibly return a related error) which only makes sense for the special case of a "exactly one child and its parents" package that `AcceptPackage()`  ensures.

Another question about RBF below.

glozow commented on src/validation.cpp:
```diff
@@ -508,6 +508,19 @@ class MemPoolAccept
         }
 
+        /** Parameters for a single transaction within a package. */
+        static ATMPArgs SingleInPackageAccept(const ATMPArgs& package_args) {
```
Ah, good point. I don't have a good reason as to why BIP125 should ever be disallowed here. Will change it to true for `SingleInPackageAccept`

mzumsande replied:
If the idea is that everything should be just as if this was a single transaction, why introduce a separate `SingleInPackageAccept` function at all, instead of just creating a `SingleAccept()` ATMPArgs set?

glozow replied:
I just figured it would be safer to have a constructor converting package-to-single ATMPArgs. There also is a difference (albeit small) - `SingleInPackageAccept` should never have `bypass_limits=true`, so I don't want the caller to have the option to set it.

glozow commented on src/validation.cpp:
```diff
@@ -1205,20 +1215,35 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
+        [](CAmount sum, auto& ws) { return sum + ws.m_modified_fees; });
+    const CFeeRate package_feerate(m_total_modified_fees, m_total_vsize);
+    TxValidationState placeholder_state;
+    if (args.m_package_feerates &&
```
Right, we always have `m_package_feerates` on when we call `AcceptMultipleTransactions`. I imagined turning it off for `testmempoolaccept` with multiple transactions but didn't actually come up with a good reason to do that. I think i will delete this condition from here.

glozow replied:
Fixed - it was a mistake to make `ATMPArgs::PackageTestAccept.m_package_feerates=true` in the original code

glozow commented:
In response to @mzumsande's (very good) questions, I have made these 2 changes:

`ATMPArgs::PackageTestAccept.m_package_feerates=false` because there is no child-with-parents restriction (it accepts completely independent transactions), which means package feerate isn't very meaningful and we shouldn't use it in `CheckFeeRate()`.

`ATMPArgs::SingleInPackageAccept.m_allow_bip125_replacement=true` because RBF should be allowed for the individual transactions.

Also addressed @LarryRuane's https://github.com/bitcoin/bitcoin/pull/24152#discussion_r836867916 which I forgot to address last time.

mzumsande commented:
Code review ACK 9bebf35e269b2a918df27708565ecd0c5bd3f116 
I am not extremely experienced with this area of code, but all changes make sense to me.
I verified via git range-diff that the only changes compared to my review from yesterday are those explained above (ATMPArgs and std::optional changes).

t-bast reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/24152/commits/9bebf35e269b2a918df27708565ecd0c5bd3f116

ariard commented:
Post-Merge Review ACK 9bebf35e

Changes since last ACK are turning `m_packages_feerates` to `false` for `PackageTestAccept` and authorizing bip 125 replacement for `SingleInPackageAccept`.

darosior commented on src/validation.cpp:
```diff
@@ -1205,20 +1215,35 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
         m_viewmempool.PackageAddTransaction(ws.m_ptx);
     }
 
+    // Transactions must meet two minimum feerates: the mempool minimum fee and min relay fee.
+    // For transactions consisting of exactly one child and its parents, it suffices to use the
+    // package feerate (total modified fees / total virtual size) to check this requirement.
+    const auto m_total_vsize = std::accumulate(workspaces.cbegin(), workspaces.cend(), int64_t{0},
```
This assumes that `AcceptMultipleTransactions` always takes proper packages and that we can safely consider the aggregated feerate of the bunch of transactions. It's not the case.
https://github.com/bitcoin/bitcoin/blob/e3c08eb620a2f317fc09fdf20969fa26f02afb91/src/validation.cpp#L555-L560

This patch demonstrates how this makes `testmempoolaccept` accept invalid transactions (by considering the package feerate for unrelated transactions):
```patch
diff --git a/test/functional/mempool_accept.py b/test/functional/mempool_accept.py
index d3961e753d..f8dd380ee5 100755
--- a/test/functional/mempool_accept.py
+++ b/test/functional/mempool_accept.py
@@ -333,6 +333,15 @@ class MempoolAcceptanceTest(BitcoinTestFramework):
             maxfeerate=0,
         )
 
+        self.log.info("Don't consider package fees between unrelated transactions")
+        txs = [
+            self.wallet.create_self_transfer(fee_rate=0, mempool_valid=False)['tx'],
+            self.wallet.create_self_transfer()['tx'],
+        ]
+        res = self.nodes[0].testmempoolaccept([tx.serialize().hex() for tx in txs])
+        # The first tx has a feerate of 0. No child pay for it. It should be rejected.
+        assert not res[0]["allowed"]
+
 
 if __name__ == '__main__':
     MempoolAcceptanceTest().main()
```

darosior reviewed (commented):
~~I think the package feerate check shouldn't be in `AcceptMultipleTransactions`, since it doesn't necessarily take a child-with-parents package.~~

~~(sorry, i know i'm late to the party)~~

EDIT: aarg. I was reviewing an older branch which didn't have the fix for `m_package_feerates` in `PackageTestAccept` (https://github.com/bitcoin/bitcoin/pull/24152#discussion_r843317502)... Sorry for the noise.

darosior reviewed (commented):
Post-merge ACK 9bebf35e269b2a918df27708565ecd0c5bd3f116
